cpp/src/arrow/vendored/ProducerConsumerQueue.h

```diff
@@ -0,0 +1,211 @@
+// Vendored from git tag v2021.02.15.00
+
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// @author Bo Hu (bhu@fb.com)
+// @author Jordan DeLong (delong.j@fb.com)
+
+#pragma once
+
+#include <atomic>
+#include <cassert>
+#include <cstdlib>
+#include <memory>
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
+
+namespace folly {
```

Comment:
I'm not sure how we want to go about this, but often we try to enclose vendored code in an additional `arrow_vendored` namespace (because when Arrow C++ is used as a library, there may be otherwise be symbol clashes with other uses of Folly by the application - which is very bad if they use a different version of Folly).

Reply:
I wrapped it in `arrow_vendored`